inlanecruising_plugin/src/inlanecruising_plugin.cpp

```diff
@@ -159,38 +235,79 @@ std::vector<PointSpeedPair> InLaneCruisingPlugin::constrain_to_time_boundary(con
   return time_bound_points;
 }
 
+int InLaneCruisingPlugin::get_nearest_point_index(const std::vector<lanelet::BasicPoint2d>& points,
+                                               const cav_msgs::VehicleState& state)
+{
+  lanelet::BasicPoint2d veh_point(state.X_pos_global, state.Y_pos_global);
+  double min_distance = std::numeric_limits<double>::max();
+  int i = 0;
+  int best_index = 0;
+  for (const auto& p : points)
+  {
+    double distance = lanelet::geometry::distance2d(p, veh_point);
+    if (distance < min_distance)
+    {
+      best_index = i;
+      min_distance = distance;
+    }
+    i++;
+  }
+
+  return best_index;
+}
+
+
 std::vector<cav_msgs::TrajectoryPlanPoint> InLaneCruisingPlugin::compose_trajectory_from_centerline(
     const std::vector<PointSpeedPair>& points, const cav_msgs::VehicleState& state)
 {
-  ROS_DEBUG_STREAM("VehicleState: "
+ ROS_DEBUG_STREAM("VehicleState: "
                    << " x: " << state.X_pos_global << " y: " << state.Y_pos_global << " yaw: " << state.orientation
                    << " speed: " << state.longitudinal_vel);
 
   ROS_DEBUG_STREAM("points size: " << points.size());
   
-  log::printDebugPerLine(points, &log::pointSpeedPairToStream);
+  //log::printDebugPerLine(points, &log::pointSpeedPairToStream);
 
-  int nearest_pt_index = getNearestPointIndex(points, state);
+  int nearest_pt_index = get_nearest_point_index(points, state);
 
   ROS_DEBUG_STREAM("NearestPtIndex: " << nearest_pt_index);
 
   std::vector<PointSpeedPair> future_points(points.begin() + nearest_pt_index + 1, points.end()); // Points in front of current vehicle position
 
+ std::vector<PointSpeedPair> back_and_future;
+  back_and_future.reserve(points.size());
+  double total_dist = 0;
+  int min_i = 0;
+  for (int i = nearest_pt_index; i >= 0; --i) { // NOTE: Do not use size_t for i type here as -- with >= 0 will result in overflow
+    min_i = i;
+    if (i == nearest_pt_index) {
+      continue;
+    }
+    total_dist += lanelet::geometry::distance2d(points[i].point, points[i-1].point);
+    if (total_dist > config_.back_distance) {
+      break;
+    }
+  }
+
   auto time_bound_points = constrain_to_time_boundary(future_points, config_.trajectory_time_length);
 
-  ROS_DEBUG_STREAM("time_bound_points: " << time_bound_points.size());
+  back_and_future.insert(back_and_future.end(), points.begin() + min_i, points.begin() + nearest_pt_index + 1);
+  back_and_future.insert(back_and_future.end(), time_bound_points.begin(), time_bound_points.end());
```

Comment:
Might be worth breaking this and the loop above into its own method so its clearer that we are adding previous points to the path.

Reply:
separated it out